diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt

```diff
@@ -36,31 +48,166 @@ class StringConcatenationRule(configRules: List<RulesConfig>) : DiktatRule(
-        assert(node.elementType == BINARY_EXPRESSION)
-        val firstChild = node.firstChildNode
-        return if (isPlusBinaryExpression(node) && firstChild.elementType == STRING_TEMPLATE) {
-            STRING_CONCATENATION.warn(configRules, emitWarn, this.isFixMode, parentNode.text, firstChild.startOffset, firstChild)
-            true
-        } else {
-            false
+    private fun circle(parentTopLevelBinaryExpr: ASTNode) {
+        val allBinaryExpressions = parentTopLevelBinaryExpr.findAllDescendantsWithSpecificType(BINARY_EXPRESSION)
+        val nodeWithBug = allBinaryExpressions.find { isDetectStringConcatenationInExpression(it) }
+
+        val bugDetected = nodeWithBug != null
+        if (bugDetected) {
+            fixBinaryExpressionWithConcatenation(nodeWithBug)
+            circle(parentTopLevelBinaryExpr)
         }
     }
 
-    @Suppress("COMMENT_WHITE_SPACE")
+    /**
+     * We can detect string concatenation by the first (left) operand in binary expression.
+     * If it is of type string - then we found string concatenation.
+     */
+    private fun isDetectStringConcatenationInExpression(node: ASTNode): Boolean {
+        assert(node.elementType == BINARY_EXPRESSION) {
+            "cannot process non binary expression in the process of detecting string concatenation"
+        }
+        val firstChild = node.firstChildNode
+        return (isPlusBinaryExpression(node) &&
+                (firstChild.elementType == STRING_TEMPLATE ||
+                        (firstChild.text.contains("toString()")) && firstChild.elementType == DOT_QUALIFIED_EXPRESSION)
+        )
+    }
+
     private fun isPlusBinaryExpression(node: ASTNode): Boolean {
         assert(node.elementType == BINARY_EXPRESSION)
-        //     binary expression
-        //    /        |        \
-        //  expr1 operationRef expr2
+        // binary expression
+        // /        |        \
+        // expr1 operationRef expr2
 
         val operationReference = node.getFirstChildWithType(OPERATION_REFERENCE)
         return operationReference
             ?.getFirstChildWithType(PLUS) != null
     }
+
+    private fun fixBinaryExpressionWithConcatenation(node: ASTNode?) {
+        val binaryExpressionNode = node?.psi as KtBinaryExpression
+        val parentNode = node.treeParent
+        val textNode = checkKtExpression(binaryExpressionNode)
+        val newNode = KotlinParser().createNode("\"$textNode\"")
+        parentNode.replaceChild(node, newNode)
+    }
+
+    private fun isPlusBinaryExpressionAndFirstElementString(binaryExpressionNode: KtBinaryExpression) =
+            (binaryExpressionNode.left is KtStringTemplateExpression) && PLUS == binaryExpressionNode.operationToken
+
+    @Suppress(
+        "TOO_LONG_FUNCTION",
+        "NESTED_BLOCK",
+        "SAY_NO_TO_VAR")
+    private fun checkKtExpression(binaryExpressionNode: KtBinaryExpression): String {
+        var lvalueText = binaryExpressionNode.left!!.text.replace("\"", "")
```

Comment:
```suggestion
        var lvalueText = binaryExpressionNode.left!!.text.trim("\"")
```
In case there are escaped quoates inside the string, or just quotes inside triple-quoted string